File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxProd = 1;
        int minProd = 1;
        int res = nums[0];
        for (int i = 0; i < nums.length; i++){
            int temp = nums[i] * maxProd;
            maxProd = Math.max(nums[i], Math.max(nums[i]*maxProd, nums[i]*minProd));
            minProd = Math.min(nums[i], Math.min(temp, nums[i]*minProd));
            res = Math.max(res, Math.max(maxProd, minProd));
        }
        return res;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        pre, suf = 1, 1
        res = -(inf)
        n = len(nums)
        for i in range(0, n):
            if (pre==0): pre = 1
            if (suf==0): suf = 1
            pre *= nums[i]
            suf *= nums[n-i-1]
            res = max(res, pre, suf)
        return res
            